tests/test_memory_message_bus.py

```diff
@@ -0,0 +1,111 @@
+        events_seen.append(event)
+        uow.add_event(EventB(id="2"))
+
+    def respond_to_b(event: EventB, uow: UnitOfWork):
+        events_seen.append(event)
+
+    uow = UnitOfWork(FakeRepositoryGateway())
+    handlers: dict[type[Event], list[Callable]] = {
+        EventA: [lambda event: respond_to_a(event, uow)],
+        EventB: [lambda event: respond_to_b(event, uow)]
+    }
+    message_bus = MemoryMessageBus(handlers)
+    
+    message_bus.handle(EventA(id="1"), uow)
+
+    assert [EventA(id="1"), EventB(id="2")] == events_seen
+
+def test_message_bus_with_single_event() -> None:
+    events_seen: list[Event] = []
+
+    def respond_to_a(event: EventA, uow: UnitOfWork):
+        events_seen.append(event)
+
+    uow = UnitOfWork(FakeRepositoryGateway())
+    handlers: dict[type[Event], list[Callable]] = {
+        EventA: [lambda event: respond_to_a(event, uow)]
+    }
+    message_bus = MemoryMessageBus(handlers)
+    
+    message_bus.handle(EventA(id="1"), uow)
+
+    assert events_seen == [EventA(id="1")]
+
+def test_message_bus_throws_error_for_event_with_no_handlers() -> None:
+    uow = UnitOfWork(FakeRepositoryGateway())
+    handlers: dict[type[Event], list[Callable]] = {}  
+    message_bus = MemoryMessageBus(handlers)
+
+    with pytest.raises(EventTypeMismatchError):
+        message_bus.handle(EventA(id="1"), uow)
+
+def test_message_bus_can_handle_multiple_handlers_for_same_event() -> None:
+    events_seen: list[Event] = []
+
+    def first_handler(event: EventA, uow: UnitOfWork):
+        events_seen.append(f"first: {event.id}")
+
+    def second_handler(event: EventA, uow: UnitOfWork):
+        events_seen.append(f"second: {event.id}")
+
+    uow = UnitOfWork(FakeRepositoryGateway())
+    handlers: dict[type[Event], list[Callable]] = {
+        EventA: [lambda event: first_handler(event, uow), lambda event: second_handler(event, uow)]
+    }
+    message_bus = MemoryMessageBus(handlers)
+
+    message_bus.handle(EventA(id="1"), uow)
+
+    assert events_seen == ["first: 1", "second: 1"]        
+
+def test_message_bus_passes_correct_event_to_handler_or_fails() -> None:
+    events_seen: list[Event] = []
+
+    def event_c_handler(event: EventC, uow: UnitOfWork):
+        print(f"Handler received event of type: {type(event)}")
+        if not isinstance(event, EventC):
+            raise EventTypeMismatchError(f"Expected EventC, got {type(event)}")
+        events_seen.append(event)
+
+    uow = UnitOfWork(FakeRepositoryGateway())
+    handlers: dict[type[Event], list[Callable]] = {
+        EventC: [lambda event: event_c_handler(event, uow)]
+    }
+    message_bus = MemoryMessageBus(handlers)
+
+    event_c = EventC(id="1")
+    other_event = OtherEvent(id="2")
+    message_bus.handle(event_c, uow)
+
+    assert events_seen == [EventC(id="1")], f"Expected [MyEvent(id='1')], but got {events_seen}"
+    with pytest.raises(EventTypeMismatchError):
+        message_bus.handle(other_event, uow)
```

Comment:
I'm a little puzzled by this test. I don't think we can verify that the event received by a handler is in fact the type in the type hint unless we check it using `is_instance` (like you've done here). But I don't think we're trying to test the code in the handlers, we want to test the message bus. In other words, I don't think it establishes much to check whether an exception was raised if you raised it in the handler.

Reply:
The concept behind this test is to verify that the message bus correctly routes events to the appropriate handler and raises an exception when an event is passed to an unexpected handler. I’m considering revising the code to clarify the exception handling in the handler. Let me know your thoughts.